website/guides/advanced/custom-aggregations-with-lua.md.erb

```diff
@@ -0,0 +1,284 @@
+  function (emit)
+    count = 0 -- initialize state by setting a global variable
+  end
+"""
+
+hooks.process = """
+  function (event, emit)
+    count = count + 1 -- increment the counter and exit
+  end
+"""
+
+timers = [{interval_seconds = 5, handler = """
+  function (emit)
+    emit {
+      metric = {
+        name = "event_counter",
+        kind = "incremental",
+        timestamp = os.date("!*t"),
+        counter = {
+          value = counter
+        }
+      }
+    }
+    counter = 0
+  end
+"""}]
+
+hooks.shutdown = """
+  function (emit)
+    emit {
+      metric = {
+        name = "event_counter",
+        kind = "incremental",
+        timestamp = os.date("!*t"),
+        counter = {
+          value = counter
+        }
+      }
+    }
+  end
+"""
+```
+
+One could plug it into a [pipeline][docs.about.concepts#pipelines] and it would work!
+
+However, this code could and should be refactored. Hold on to the next section to see way in which it could be done.
+
+## Reduce Duplication
+
+A bird's-eye view of the transform definition reveals that the timer handler and the shutdown hook are almost
+identical. It is possible make the config more [DRY][urls.dry_code] by extracting creation of the counter into
+a dedicated function. Such a function can be placed into the [source][docs.reference.transforms.lua#source]
+section of the config:
+
+```toml
+source = """
+  function make_counter(value)
+    return metric = {
+      name = "event_counter",
+      kind = "incremental",
+      timestamp = os.date("!*t"),
+      counter = {
+        value = value
+      }
+    }
+  end
+"""
+```
+
+and then adjusting the timer handler
+
+```toml
+timers = [{interval_seconds = 5, handler = """
+  function (emit)
+    emit(make_counter(counter))
+    counter = 0
+  end
+"""}]
+```
+
+and the `shutdown` hook:
+
+```toml
+hooks.shutdown = """
+  function (emit)
+    emit(make_counter(counter))
+  end
+"""
+```
+
+## Keep All Code Together
+
+The new config looks tidier, but in order to make it more readable, it is also possible to gather implementations of
+all functions into the `source` section, resulting in the following component declaration:
+
+```toml title="vector.toml"
+[transforms.aggregator]
+type = "lua"
+version = "2"
+inputs = [] # add names of the input components here
+hooks.init = "init"
+hooks.process = "process"
+hooks.shutdown = "shutdown"
+timers = [{interval_seconds = 5, handler = "timer_handler"}]
+
+source = """
+  function init()
+    count = 0
+  end
+
+  function process()
+    count = count + 1
+  end
+
+  function timer_handler(emit)
+    emit(make_counter(counter))
+    counter = 0
+  end
+
+  function shutdown(emit)
+    emit(make_counter(counter))
+  end
+
+  function make_counter(value)
+    return metric = {
+      name = "event_counter",
+      kind = "incremental",
+      timestamp = os.date("!*t"),
+      counter = {
+        value = value
+      }
+    }
+  end
+"""
+```
+
+## A Loadable Module
```

Comment:
I really like the progressive nature of this guide and how you clearly point out the advantages of each approach.